File: 239-sliding-window-maximum/sliding-window-maximum.py
# Brute Force Code & Optimal Code
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # max heap implemented using negative values
        max_heap = []
        # list to store the maximum element of each sliding window
        result = []
        # total number of element in the array
        n = len(nums)
        # traverse the array
        for i in range(n):
            # remove element from the heap that are no longers inside the current sliding windows 
            while (max_heap and max_heap[0][1] <= i - k):
                heapq.heappop(max_heap)
            # insert the current element along with its index into the max heap
            heapq.heappush(max_heap, (-nums[i], i))
            # once the first window of size k is formed the heap top element is the maximum values
            if i >= k - 1:
                result.append(-max_heap[0][0])
        # return the maximum values for every sliding winodw
        return result 

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 239-sliding-window-maximum/sliding-window-maximum.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        // max heap implemented using negative values
        priority_queue<pair<int, int>> max_heap;
        // list to store the maximum element of each sliding window
        vector<int> result;
        // total number of elements in the array
        int n = nums.size();
        // traverse the array
        for (int i = 0; i < n; i++) {
            // remove elements from the heap that are no longer inside the current sliding window
            while (!max_heap.empty() &&
                   max_heap.top().second <= i - k) {
                max_heap.pop();
            }
            // insert the current element along with its index into the max heap
            max_heap.push({nums[i], i});
            // once the first window of size k is formed the heap top element is the maximum value
            if (i >= k - 1) {
                result.push_back(max_heap.top().first);
            }
        }
        // return the maximum values for every sliding window
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)